### PR TITLE
Improve the yaml minimization algorithm (again).

### PR DESCRIPTION
**Pull request by johnynek**
@ianoc-stripe in #84 found an example of the minimization algorithm deleting entries.

I have reworked the algorithm again to be more principled, but also exponential in complexity. The exponential part only runs for sets of dependencies that share a common prefix in maven group and in artifact, so in practice that is small enough that it is very fast. But if you have 100 modules in a single artifact each with different versions, this algorithm might fall over.

It is only used in format now, so even if we hit such a corner case, users can work around by not formatting (which triggers the minimization).

### CODE REVIEW

**Comment by johnynek** [comment]
@ianoc-stripe @ianoc take a look?

File: test/scala/com/github/johnynek/bazel_deps/ParseTestCasesTest.scala [hunk 1]
```diff
@@ -98,4 +98,42 @@ dependencies:
...
+    chill:
+      lang: scala
+      version: "0.8.4"
+    chill-algebird:
+      exports:\u0020
+        - "com.twitter:chill"
+      lang: scala
+      version: "0.8.4"
+    chill-scrooge:
+      exports:\u0020
```
Comment (johnynek) [inline test/scala/com/github/johnynek/bazel_deps/ParseTestCasesTest.scala]: these suck, but that is a bigger issue: remove trailing whitespace in the generated docs.

Comment (ianoc) [reply test/scala/com/github/johnynek/bazel_deps/ParseTestCasesTest.scala]: Yeah not a big deal, could do it as post-processing if its hard to fit into the code here too i imagine?  (In one of the projects i have i'm just doing a regex against the file after the formatter runs to strip this)

File: src/scala/com/github/johnynek/bazel_deps/DepsModel.scala [hunk 2]
```diff
@@ -668,6 +621,131 @@ case class Dependencies(toMap: Map[MavenGroup, Map[ArtifactOrProject, ProjectRec
 object Dependencies {
   def empty: Dependencies = Dependencies(Map.empty[MavenGroup, Map[ArtifactOrProject, ProjectRecord]])
 
+  /**
+   * Combine as many ProjectRecords as possible into a result
+   */
+  def normalize(candidates0: List[(ArtifactOrProject, ProjectRecord)]): List[(ArtifactOrProject, ProjectRecord)] = {
+    type AP= (ArtifactOrProject, ProjectRecord)
```
Comment (ianoc) [inline src/scala/com/github/johnynek/bazel_deps/DepsModel.scala]: Random style nit, but i presume should be a space after the `AP` ?

File: test/scala/com/github/johnynek/bazel_deps/ParseGeneratedDocTest.scala [hunk 3]
```diff
@@ -20,7 +20,7 @@ class ParseGeneratedDocTest extends FunSuite {
 
     forAll(genList) { lp =>
       val output = Dependencies.normalize(lp)
-      assert(lp.size <= output.size)
+      assert(lp.size >= output.size)
```
Comment (ianoc) [inline test/scala/com/github/johnynek/bazel_deps/ParseGeneratedDocTest.scala]: Is it bad that this didn't fail in tests if the law was incorrect? (build looks green, but obviously scalacheck style so might not mean much?)

**Comment by ianoc** [comment]
Don't really know the code super well to be able to review in a great fashion. But it looks pretty sane and good test coverage so i'd be 👍

File: test/scala/com/github/johnynek/bazel_deps/ParseGeneratedDocTest.scala (hunk 3 above)
Comment (johnynek) [reply test/scala/com/github/johnynek/bazel_deps/ParseGeneratedDocTest.scala]: yeah, I think the scalacheck is *very* rarely hitting cases where minimization are running, so basically only the handwritten cases are exercising.

It would be nice to make a scala-check `Gen` that makes minimization more likely.

File: src/scala/com/github/johnynek/bazel_deps/DepsModel.scala (hunk 2 above)
Comment (johnynek) [reply src/scala/com/github/johnynek/bazel_deps/DepsModel.scala]: I'll fix that. Thanks for catching!

**Comment by ianoc** [comment]
👍

File: test/scala/com/github/johnynek/bazel_deps/ParseTestCasesTest.scala (hunk 1 above)
Comment (johnynek) [reply test/scala/com/github/johnynek/bazel_deps/ParseTestCasesTest.scala]: yeah, I want to fix this. I'll do it in a separate issue.

There is a new method in paiges which should do this (`.renderTrim`) but I found a bug in it when I tried to use it.
